### PR TITLE
Support `TypeAliasType`

### PR DESCRIPTION
**Pull request by hamdanal**
Builds on top of and supersedes #16644

### CODE REVIEW

File: mypy/semanal.py [hunk 1]
```diff
@@ -2007,34 +2007,36 @@ def analyze_class_typevar_declaration(self, base: Type) -> tuple[TypeVarLikeList
...
                 return None
             return self.analyze_unbound_tvar_impl(t)
         return None
 
     def analyze_unbound_tvar_impl(
-        self, t: UnboundType, allow_tvt: bool = False
+        self, t: UnboundType, is_unpacked: bool = False, is_typealias_param: bool = False
     ) -> tuple[str, TypeVarLikeExpr] | None:
+        if is_unpacked and is_typealias_param:
+            return None  # This should be unreachable
```
Comment (JelleZijlstra) [inline mypy/semanal.py:2025]: Maybe add an assertion in that case?

File: test-data/unit/check-generics.test [hunk 2]
```diff
@@ -3124,8 +3124,8 @@ def dec(f: Callable[P, T]) -> Callable[P, List[T]]: ...
 def id(x: U) -> U: ...
 def either(x: U, y: U) -> U: ...
 def pair(x: U, y: V) -> Tuple[U, V]: ...
-reveal_type(dec(id))  # N: Revealed type is "def [T] (x: T`2) -> builtins.list[T`2]"
-reveal_type(dec(either))  # N: Revealed type is "def [T] (x: T`4, y: T`4) -> builtins.list[T`4]"
+reveal_type(dec(id))  # N: Revealed type is "def [T] (x: T`3) -> builtins.list[T`3]"
```
Comment (JelleZijlstra) [inline test-data/unit/check-generics.test:3127]: Do you know why this changed?

File: test-data/unit/check-type-aliases.test [hunk 3]
```diff
@@ -1065,3 +1065,99 @@ def eval(e: Expr) -> int:
...
+[case testTypeAliasTypeInvalid]
+from typing_extensions import TypeAliasType
+
+TestType = TypeAliasType("T", int)  # E: String argument 1 "T" to TypeAliasType(...) does not match variable name "TestType"
+
+T1 = T2 = TypeAliasType("T", int)
+t1: T1  # E: Variable "__main__.T1" is not valid as a type \
+        # N: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
+
+T3 = TypeAliasType("T3", -1)
```
Comment (JelleZijlstra) [inline test-data/unit/check-type-aliases.test:1088]: Shouldn't this line produce an error?

File: test-data/unit/check-type-aliases.test [hunk 4]
```diff
@@ -1065,3 +1065,99 @@ def eval(e: Expr) -> int:
...
+from typing import Callable, Dict, TypeVar, Tuple, Unpack
+
+K = TypeVar('K')
+V = TypeVar('V')
+Ts = TypeVarTuple("Ts")
+Ts1 = TypeVarTuple("Ts1")
+P = ParamSpec("P")
+
+TestType = TypeAliasType("TestType", Dict[K, V], type_params=(K, V))
+x: TestType[int, str] = {1: 'a'}
```
Comment (JelleZijlstra) [inline test-data/unit/check-type-aliases.test:1104]: Test that you get an error with the wrong number of type parameters (e.g., `TestType[int]`).

File: test-data/unit/check-type-aliases.test [hunk 5]
```diff
@@ -1065,3 +1065,99 @@ def eval(e: Expr) -> int:
...
+V = TypeVar('V')
+Ts = TypeVarTuple("Ts")
+Ts1 = TypeVarTuple("Ts1")
+P = ParamSpec("P")
+
+Ta1 = TypeAliasType("Ta1", int, type_params=K)  # E: Tuple literal expected as the type_params argument to TypeAliasType
+
+Ta2 = TypeAliasType("Ta2", int, type_params=(None,))  # E: Free type variable expected in type_params argument to TypeAliasType
+
+Ta3 = TypeAliasType("Ta3", Dict[K, V], type_params=(V,))  # E: Can't use bound type variable "K" to define generic alias
```
Comment (JelleZijlstra) [inline test-data/unit/check-type-aliases.test:1142]: Not a very clear error. Can this instead say "Type variable K is not included in type_params" or something similar?

File: test-data/unit/check-type-aliases.test [hunk 6]
```diff
@@ -1065,3 +1065,99 @@ def eval(e: Expr) -> int:
...
+partially_generic1: Ta3[int] = {"a": 1}
+reveal_type(partially_generic1)  # N: Revealed type is "builtins.dict[Any, builtins.int]"
+partially_generic2: Ta3[int] = {1: "a"}  # E: Dict entry 0 has incompatible type "int": "str"; expected "Any": "int"
+
+Ta4 = TypeAliasType("Ta4", Tuple[Unpack[Ts]], type_params=(Ts, Ts1))  # E: Can only use one type var tuple in type_params argument to TypeAliasType
+
+Ta5 = TypeAliasType("Ta5", Dict)  # Unlike old style aliases, this is not generic
```
Comment (JelleZijlstra) [inline test-data/unit/check-type-aliases.test:1149]: Add a test that demonstrates that this is equivalent to Dict[Any, Any]

File: test-data/unit/check-type-aliases.test (hunk 5 above)
Comment (hamdanal) [reply test-data/unit/check-type-aliases.test:1142]: I like this suggestion. Unless you have an alternative that would also cover the same case but for aliases written with the `type` keyword, I'll go with the your current suggestion.

File: test-data/unit/check-generics.test (hunk 2 above)
Comment (hamdanal) [reply test-data/unit/check-generics.test:3127]: I have no idea to be honest. They changed in the original PR by sobolevn and I didn't know why. I'll do some more digging.

Comment (JelleZijlstra) [reply test-data/unit/check-generics.test:3127]: Thanks. Not too important if it's hard to figure out, I assume these numbers can change based on all kinds of factors.

File: test-data/unit/check-type-aliases.test [hunk 7]
```diff
@@ -1065,3 +1065,125 @@ def eval(e: Expr) -> int:
...
+                                     # E: Type variable "T" is not included in type_params
+unbound_ps_alias: Ta7[[int], str]  # E: Bracketed expression "[...]" is not valid as a type \
+                                   # N: Did you mean "List[...]"? \
+                                   # E: Bad number of arguments for type alias, expected 0, given 2
+reveal_type(unbound_ps_alias)  # N: Revealed type is "__main__.G[Any, Any]"
+
+# TODO this does not work yet, it should report unbound P
+# Ta8 = TypeAliasType("Ta8", Callable[P, int])
+# unbound_ps_alias: Ta8[int]
+# reveal_type(unbound_ps_alias)
```
Comment (hamdanal) [inline test-data/unit/check-type-aliases.test:1175]: `Callable` aliases seem to be special cased some how. This does not report the unbound type for some reason. I'll need to investigate further.

Comment (JelleZijlstra) [reply test-data/unit/check-type-aliases.test:1175]: I'm fine with leaving this to a future PR if necessary.

File: mypy/semanal.py [hunk 8]
```diff
@@ -3764,6 +3785,75 @@ def check_and_set_up_type_alias(self, s: AssignmentStmt) -> bool:
...
+            if not isinstance(base, UnboundType):
+                continue
+
+            tag = self.track_incomplete_refs()
+            tvar = self.analyze_unbound_tvar_impl(base, is_typealias_param=True)
+            if tvar:
+                if isinstance(tvar[1], TypeVarTupleExpr):
+                    if have_type_var_tuple:
+                        self.fail(
+                            "Can only use one type var tuple in type_params argument to TypeAliasType",
```
Comment (JelleZijlstra) [inline mypy/semanal.py:3832]:
```suggestion
                            "Can only use one TypeVarTuple in type_params argument to TypeAliasType",
```

**Review by JelleZijlstra — COMMENTED** [review]
Almost there, added a suggestion for improving wording

File: mypy/typeanal.py [hunk 9]
```diff
@@ -1427,6 +1452,16 @@ def analyze_callable_args(
...
+                if self.defining_alias and self.has_type_params:
+                    tvar_likes = self.find_type_var_likes(arg)
+                    for name, tvar_expr in tvar_likes:
+                        if (name, tvar_expr) not in self.allowed_alias_tvars:
+                            self.fail(
+                                f'Type variable "{name}" is not included in type_params',
+                                arglist,
+                                code=codes.VALID_TYPE,
+                            )
+                            return None
```
Comment (hamdanal) [inline mypy/typeanal.py:1464]: Added this to handle `Callable[[Unpack[Ts]], Ret]` when `Ts` is not declared in the type parameters. It was causing a crash.

File: test-data/unit/check-type-aliases.test (hunk 7 above)
Comment (hamdanal) [reply test-data/unit/check-type-aliases.test]: This is fixed now.

**Comment by JelleZijlstra** [comment]
Tests are failing, I think because the `*Ts` syntax is 3.12 only. Maybe split that part of the test into its own case and make it run `--python-version 3.12`.

File: test-data/unit/check-type-aliases.test (hunk 7 above)
Comment (JelleZijlstra) [reply test-data/unit/check-type-aliases.test]: Great, thanks!

**Comment by hamdanal** [comment]
> Tests are failing, I think because the `*Ts` syntax is 3.12 only. Maybe split that part of the test into its own case and make it run `--python-version 3.12`.

This is already the case. Maybe I’ll need to move the test to a Python 3.12+ test file?

File: mypy/typeanal.py (hunk 9 above)
Comment (hamdanal) [reply mypy/typeanal.py]: I think I'll revert this part, I don't think it is the correct solution here.

Comment (hamdanal) [reply mypy/typeanal.py]: OK so looks like this already crashes on master with old aliases https://mypy-play.net/?mypy=master&python=3.12&gist=744b559e656f6f35cd5762bddf979038. I'll open an issue to track it.

File: mypy/semanal.py [hunk 10]
```diff
@@ -2018,34 +2018,36 @@ def analyze_class_typevar_declaration(self, base: Type) -> tuple[TypeVarLikeList
...
                 return None
             return self.analyze_unbound_tvar_impl(t)
         return None
 
     def analyze_unbound_tvar_impl(
-        self, t: UnboundType, allow_tvt: bool = False
+        self, t: UnboundType, is_unpacked: bool = False, is_typealias_param: bool = False
     ) -> tuple[str, TypeVarLikeExpr] | None:
+        if is_unpacked and is_typealias_param:
+            assert False, "Unreachable"
```
Comment (JelleZijlstra) [inline mypy/semanal.py:2036]:
```suggestion
        assert not unpacked or not is_typelias_param, "Mutually exclusive conditions"
```

File: mypy/semanal.py [hunk 11]
```diff
@@ -2018,34 +2018,35 @@ def analyze_class_typevar_declaration(self, base: Type) -> tuple[TypeVarLikeList
...
+                    return self.analyze_unbound_tvar_impl(inner_t, is_unpacked=True)
                 return None
             return self.analyze_unbound_tvar_impl(t)
         return None
 
     def analyze_unbound_tvar_impl(
-        self, t: UnboundType, allow_tvt: bool = False
+        self, t: UnboundType, is_unpacked: bool = False, is_typealias_param: bool = False
     ) -> tuple[str, TypeVarLikeExpr] | None:
+        assert not unpacked or not is_typelias_param, "Mutually exclusive conditions"
```
Comment (JelleZijlstra) [inline mypy/semanal.py:2035]:
```suggestion
        assert not is_unpacked or not is_typelias_param, "Mutually exclusive conditions"
```

File: mypy/semanal.py [hunk 12]
```diff
@@ -2018,34 +2018,35 @@ def analyze_class_typevar_declaration(self, base: Type) -> tuple[TypeVarLikeList
...
+                    return self.analyze_unbound_tvar_impl(inner_t, is_unpacked=True)
                 return None
             return self.analyze_unbound_tvar_impl(t)
         return None
 
     def analyze_unbound_tvar_impl(
-        self, t: UnboundType, allow_tvt: bool = False
+        self, t: UnboundType, is_unpacked: bool = False, is_typealias_param: bool = False
     ) -> tuple[str, TypeVarLikeExpr] | None:
+        assert not is_unpacked or not is_typelias_param, "Mutually exclusive conditions"
```
Comment (JelleZijlstra) [inline mypy/semanal.py:2035]:
```suggestion
        assert not is_unpacked or not is_typealias_param, "Mutually exclusive conditions"
```
I can spell

**Pull request merged**